### ISSUE TITLE
Improve feedback to the user during startup

### ISSUE THREAD

**Issue by gbraad** [minishift/minishift#485]
As part of an improved user experience, the user should be told what defaults has been chosen for him, or which values he has given, to make the starting process easier to understand.

For instance,
  * initial/default values can be conveyed to the user
  * Misinterpreted values become visible (use as 4 or 4096, when the user expects something to be in GB?)
  * Progress tracking, which can indicate a point of failure/slowdown to be user, without having to resort to the libmachine logs.

**Comment by bronhaim** [comment]
- The defaults are presented as part of the help doc which is enough imo. 
- start --memory flag accepts only int value for amount of mb - so --memory 4GB doesn't exist.
- The progress tracking - do you mean some kind of ./minishift events watch command ? not sure its really needed. there is status command which is good enough. maybe we can improve the output there.

Maybe this issue should be about improving "start" output. currently its the content below. maybe it should print also vm's details - how much memory, cpus it uses for example.

Although this info is quite easy to reach using virsh dumpxml.. let me know if you want that


--- snip
Starting local OpenShift cluster using 'kvm' hypervisor...
Provisioning OpenShift via '/home/ybronhaim/.minishift/cache/oc/v1.4.1/oc [cluster up --use-existing-config --host-config-dir /var/lib/minishift/openshift.local.config --host-data-dir /var/lib/minishift/hostdata --host-volumes-dir /var/lib/minishift/openshift.local.volumes]'
-- Checking OpenShift client ... OK
-- Checking Docker client ... OK
-- Checking Docker version ... OK
-- Checking for existing OpenShift container ... 
   Deleted existing OpenShift container
-- Checking for openshift/origin:v1.4.1 image ... OK
-- Checking Docker daemon configuration ... OK
-- Checking for available ports ... OK
-- Checking type of volume mount ... 
   Using Docker shared volumes for OpenShift volumes
-- Creating host directories ... OK
-- Finding server IP ... 
   Using 192.168.42.143 as the server IP
-- Starting OpenShift container ... 
   Starting OpenShift using container 'origin'
   Waiting for API server to start listening
   OpenShift server started
-- Removing temporary directory ... OK
-- Server Information ... 
   OpenShift server started.
   The server is accessible via web console at:
       https://192.168.42.143:8443

   To login as administrator:
       oc login -u system:admin

**Comment by gbraad** [comment]
@bronhaim a principle of HCI is to have the user be in control, and if defaults are chosen, they should be given or shown. At the moment, when `minishift start` is done, no feedback is given. this leaves people guessing. I understand that document is a way to provide this information, but it is often not up-to-date, or people just do not consult the document. ;-)

The progress tracking is more along the lines of #480. When we automount hostfolders, this takes time (due to contacting the VM over SSH, the actual mount process, and the roundtrip of the status and error codes). This can take up seconds or more, which without any feedback, leaves the user guessing of progress. Again this is part of good HCI. The idea is along lines of what OpenShift itself does. We could show:

```
Starting lovsl OpenShift cluster
-- Provision hostname
-- Mounting hostfolders
    Mounting 'Disks': //remotemachine/Disks/ ... OK
-- Provision OpenShift via
```
Because between `Starting local OpenShift cluster` and `Provisioning OpenShift` more happens... 

Actual VM usage? Well, this is not something we consider... but do you think this is useful? On KVM, Virtualbox, for instance, this should be static. And I do not mean with ballooning involved, but for Hyper-V this can be scaled dynamically, even beyond the initial provisioning. In this case, this information could be helpful. Not sure if libmachine covers this, but for Hyper-V this information can be retrieved with PowerShell.

**Comment by praveenkumar** [comment]
During testing with openstack CI @agajdosi hit another situation when VM is not able to mount the disk and it failing without any feedback. Something we can add in the part of feedback process where tell user that mount didn't happen.

**Comment by gbraad** [comment]
Needed pre-flight checks:
  * does the minishift instance have an IP address?
  * can a remote IP address be reached? (check connectivity)
  * can a remote host be resolved? (check dns)
  * is persistent disk mounted
  * ...

**Comment by gbraad** [comment]
Started with defining several categories of tasks:
  * preflight checks before start of VM
     * has a virtual switch been selected (hyper-v)
     * correct size specs set
  * preflight check after start of VM  
     happens before doing `oc cluster up`
     * usable IP address (IPv)
     * network connectivity
     * storage mounted
  * progress tracking during clusterup
     * provisioning steps


Example of a current test (only warn with FAIL):
```
-- Checking host '8.8.8.8' reachable from VM ... FAIL
-- Checking host 'www.redhat.com' reachable from VM ... FAIL
-- Checking OpenShift client ... OK
-- Checking Docker client ... OK
-- Checking Docker version ... OK
-- Checking for existing OpenShift container ... OK
-- Checking for openshift/origin:v1.5.1 image ... 
   Pulling image openshift/origin:v1.5.1
FAIL
   Error: error pulling Docker image openshift/origin:v1.5.1
   Caused By:
     Error: Network timed out while trying to connect to https://index.docker.io/v1/repositories/openshift/origin/images. You may want to check your internet connection or if you are behind a proxy.
OpenShift provisioning failed. origin container failed to start.
```

**Comment by gbraad** [comment]
If issues occur during cluster.StartHost, such as:

```
(minishift) DBG | Running command: /usr/local/bin/docker-machine-driver-xhyve [xhyve -A -U F4BB3F79-AB4E-4708-95CA-E32FBFCDEFD3 -c 2 -m 2048M -l com1,autopty -s 0:0,hostbridge -s 31,lpc -s 2:0,virtio-net -s 3:0,ahci-cd,/Users/gbraad/.minishift/machines/minishift/boot2docker.iso -s 4:0,ahci-hd,/dev/rdisk5 -f kexec,/Users/gbraad/.minishift/machines/minishift/vmlinuz64,/Users/gbraad/.minishift/machines/minishift/initrd.img,loglevel=3 user=docker console=ttyS0 console=tty0 noembed nomodeset norestore waitusb=10:LABEL=boot2docker-data base -M]
(minishift) DBG | Converted MAC address: 2:51:8:22:87:a6
(minishift) Starting minishift...
(minishift) DBG | [xhyve -A -U F4BB3F79-AB4E-4708-95CA-E32FBFCDEFD3 -c 2 -m 2048M -l com1,autopty -s 0:0,hostbridge -s 31,lpc -s 2:0,virtio-net -s 3:0,ahci-cd,/Users/gbraad/.minishift/machines/minishift/boot2docker.iso -s 4:0,ahci-hd,/dev/rdisk5 -f kexec,/Users/gbraad/.minishift/machines/minishift/vmlinuz64,/Users/gbraad/.minishift/machines/minishift/initrd.img,loglevel=3 user=docker console=ttyS0 console=tty0 noembed nomodeset norestore waitusb=10:LABEL=boot2docker-data base -F /Users/gbraad/.minishift/machines/minishift/minishift.pid]
(minishift) Waiting for VM to come online...
(minishift) DBG | 192.168.64.2
(minishift) DBG | IP found in DHCP lease table: 192.168.64.2
(minishift) DBG | Got an ip: 192.168.64.2
(minishift) DBG | Getting to WaitForSSH function...
(minishift) DBG | Error getting ssh command 'exit 0' : Host is not running
(minishift) DBG | operation not supported by device
(minishift) Waiting on a pseudo-terminal to be ready... done
(minishift) Hook up your terminal emulator to /dev/ttys003 in order to connect to your VM
(minishift) DBG | Getting to WaitForSSH function...
(minishift) DBG | Using SSH client type: external
(minishift) DBG | Using SSH private key: /Users/gbraad/.minishift/machines/minishift/id_rsa (-rw-------)
(minishift) DBG | &{[-F /dev/null -o PasswordAuthentication=no -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o LogLevel=quiet -o ConnectionAttempts=3 -o ConnectTimeout=10 -o ControlMaster=no -o ControlPath=none docker@192.168.64.2 -o IdentitiesOnly=yes -i /Users/gbraad/.minishift/machines/minishift/id_rsa -p 22] /usr/bin/ssh <nil>}
(minishift) DBG | About to run SSH command:
(minishift) DBG | exit 0
(minishift) DBG | rdmsr to register 0x34 on vcpu 1
(minishift) DBG | SSH cmd err, output: exit status 255: 
(minishift) DBG | Error getting ssh command 'exit 0' : Something went wrong running an SSH command!
(minishift) DBG | command : exit 0
(minishift) DBG | err     : exit status 255
(minishift) DBG | output  : 
(minishift) DBG | 
(minishift) DBG | Getting to WaitForSSH function...
(minishift) DBG | Using SSH client type: external
(minishift) DBG | Using SSH private key: /Users/gbraad/.minishift/machines/minishift/id_rsa (-rw-------)
(minishift) DBG | &{[-F /dev/null -o PasswordAuthentication=no -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o LogLevel=quiet -o ConnectionAttempts=3 -o ConnectTimeout=10 -o ControlMaster=no -o ControlPath=none docker@192.168.64.2 -o IdentitiesOnly=yes -i /Users/gbraad/.minishift/machines/minishift/id_rsa -p 22] /usr/bin/ssh <nil>}
(minishift) DBG | About to run SSH command:
(minishift) DBG | exit 0
(minishift) DBG | SSH cmd err, output: exit status 255: 
(minishift) DBG | Error getting ssh command 'exit 0' : Something went wrong running an SSH command!
(minishift) DBG | command : exit 0
(minishift) DBG | err     : exit status 255
(minishift) DBG | output  : 
(minishift) DBG | 
```

it is helpful to provide an indication of progress. This could either my textual representation or a spinner. Looking into what is better suited.

**Comment by gbraad** [comment]
Including a spinner for the step of `-- Creating VM ...` might help also. Currently doing the following
```
gbraad-mbp:minishift gbraad$ minishift start
-- Installing default add-ons ... OK
-- Checking
-- Starting local OpenShift cluster using 'xhyve' hypervisor
Downloading ISO 'https://github.com/minishift/minishift-b2d-iso/releases/download/v1.0.2/minishift-b2d.iso'
 40.00 MiB / 40.00 MiB [==========================================================================================] 100.00% 0s
   Creating Minishift VM ... OK
Checking for IP ...
-- Checking host '8.8.8.8' reachable from VM ... FAIL
-- Checking host 'www.redhat.com' reachable from VM ... FAIL
-- Checking storage disk mounted ... OK
Downloading OpenShift binary 'oc' version 'v1.5.1'
 18.94 MiB / 18.94 MiB [==========================================================================================] 100.00% 0s
-- Checking OpenShift client ... OK
-- Checking Docker client ... OK
-- Checking Docker version ... OK
-- Checking for existing OpenShift container ... OK
-- Checking for openshift/origin:v1.5.1 image ...
   Pulling image openshift/origin:v1.5.1
FAIL
   Error: error pulling Docker image openshift/origin:v1.5.1
   Caused By:
     Error: Network timed out while trying to connect to https://index.docker.io/v1/repositories/openshift/origin/images. You
may want to check your internet connection or if you are behind a proxy.
OpenShift provisioning failed. origin container failed to start.
```

Note: network failures occur because I am connected to the VPN (a common problem). However, the issue I see is that `Creating VM` is a step that can take just a few seconds or more... some indication of progress is needed.

**Comment by gbraad** [comment]
Looking into the checks before start, such as driver check.

**Comment by gbraad** [comment]
```
-- Checking if 'kvm' driver installed ... OK
-- Starting local OpenShift cluster using 'kvm' hypervisor ...
-- Instance will be configured with ...
   Memory:    2048
   vCPUs :    2
   Disksize:  20000
-- Creating Minishift VM ... \ 
```

**Comment by praveenkumar** [comment]
> Memory:    2048
   vCPUs :    2
   Disksize:  20000

Instead of 20000 can we have it like `20G` or this is directly what we got from the machine config?

**Comment by gbraad** [comment]
@praveenkumar Currently trying to get everything in there, and then abstract and use a generic approach (to prevent duplication of code). But these are improvements that are certainly possible. We already have a function called `calculateDiskSizeInMB` for which we can also create a similar function `calculateDiskSizeInGB`

**Comment by gbraad** [comment]
After dealing with macOS and Linux (easy checks), I will now see what I can do to add checks for Windows Hyper-V

**Comment by LalatenduMohanty** [comment]
> -- Checking host '8.8.8.8' reachable from VM ... FAIL -- Checking host 'www.redhat.com' reachable from VM ... FAIL

+1 for the intention behind the checks. However I feel we should not just ping to "redhat.com"  we need something little more generic and elegant.  Is there any other method to do these checks?

**Comment by gbraad** [comment]
same answer, `8.8.8.8` is used to test network connectivity, while `redhat.com` is used to test the connectivity including DNS resolution. However, I want to use `index.docker.io`, but since this is hosted on EC2 it is not possible to use this for pinging. This is the actual location that is used to download an image from. The actual URL is configurable...  but suggestion what else could be used?

**Comment by gbraad** [comment]
Re-using the current version of go-units results in:
```
-- Instance will be configured with ...
   Memory:    2.048 GB
   vCPUs :    2
   Disksize:  40 GB
```
Note: current version does not have the function `HumanSizeWithPrecision`

**Comment by jorgemoralespou** [comment]
@gbraad @LalatenduMohanty I have found this to be a useful way to validate internet connectivity and dns resolution. I use it in (oc-cluster)[https://github.com/openshift-evangelists/oc-cluster-wrapper/blob/master/oc-cluster#L202]

````
curl -s -m 5 http://google.com > /dev/null 2>&1
````

**Comment by gbraad** [comment]
@jorgemoralespou The question @LalatenduMohanty actually asked is what to use as endpoint. Using `redhat.com` seems very partial... while `google.com` does not work everywhere in the world (like China). We search for something neutral. The actual test is similar to what you specified, just a shorter timeout.

**Comment by gbraad** [comment]
Additional preflight check:
  * check for diskspace on the persistent disk

**Comment by gbraad** [comment]
On Windows you can not query for the existence of Hyper-V without Administrator rights. In this I am trying to see what can be done with just the commandlets, eg. `Get-VMSwitch`, being available

**Comment by gbraad** [comment]
Docker machine's driver plugin has the following functionality as part of `powershell.go`. Let;s see if we can call this earlier:
```
func hypervAvailable() error {
	stdout, err := cmdOut("@(Get-Command Get-VM).ModuleName")
	if err != nil {
		return err
	}

	resp := parseLines(stdout)
	if resp[0] != "Hyper-V" {
		return ErrNotInstalled
	}

	return nil
}
```

**Comment by gbraad** [comment]
This is part of the `PreCreateCheck`, but this also deals with the b2d image?

**Comment by gbraad** [comment]
Will implement implementation code to: 

Treat as
  * Warning (or Info - no action needed)
  * Error (will exit)
  * Confirmation (will need a confirmation if test failed, else exit as Error)

General execution code that will check for `bool` state, and based on the `TreatAs` will fail or pass accordingly. This will also accept a possible override by config option for each individual test. Note: two implementations are needed (without and with driver as parameter), due to the Pre- and Post- create checks.

**Comment by gbraad** [comment]
Adding check to see if VM is reachable from the host. This is a common situation that occurs when a VPN is used, but also happen on xhyve for unknown reasons.

**Comment by gbraad** [comment]
It seems "CreateHost" actually deals with the provisioning (and SSH) before I am able to check the connecitivity to the VM. Will see what else can be done to do this earlier...

**Comment by gbraad** [comment]
```
-- Checking if 'kvm' driver installed ... SKIP
-- Starting local OpenShift cluster using 'kvm' hypervisor ...
-- Checking for IP ... OK
-- Checking host '8.8.8.8' reachable from VM ... OK
-- Checking host 'www.redhat.com' reachable from VM ... OK
-- Checking storage disk mounted ... OK
-- Checking storage disk usage ... 100% FAIL
   Instance does not have enough storage available
```

**Comment by hferentschik** [comment]
> -- Checking host '8.8.8.8' reachable from VM ... OK       <- default treated as warning; configurable host
> -- Checking host 'minishift.io' reachable from VM ... OK     <- default treated as warning; configurable host

I have to say I am not a big fan of these two. What are we trying to mitigate here? Obviously we cannot have them as generic tests and if you make it configurable, how would that look like. Would that not add unnecessary complexity?

**Comment by gbraad** [comment]
Connectivity issues where we are unable to pull images from internet:

Caused by:
  * User having set an internet virtual switch on Hyper-V
  * User only having host-only network on VirtualBox
  * User having non-default bridged network on KVM / Libvirt
  * Startup issues with xhyve (sometimes I had no outside network, dhcp lease issue?)

These are already treated as warning anyways...

**Comment by gbraad** [comment]
Will have to change some strings... integration failed for the following reason:

In `start.go`, when using:

```go
	print("-- Starting Minishift VM ... ")
```

the messages were logged as `stderr`, while using:

```go
	fmt.Printf("-- Starting Minishift VM ... ")
```

seems to be OK. 😕

**Comment by gbraad** [comment]
@LalatenduMohanty @praveenkumar your review of the PR is requested.

**Comment by gbraad** [comment]
Moved feedback for cluster up flags to this task. Needed code is here: https://gist.github.com/gbraad/0d6812135ce04954b187e850c07d0059

**Comment by gbraad** [comment]
```
-- Checking if 'xhyve' driver installed ... 
   Driver is available at /usr/local/bin/docker-machine-driver-xhyve
   Checking for setuid bit ... OK
-- Starting local OpenShift cluster using 'xhyve' hypervisor ...
-- Instance will be configured with ...
   Memory:    2 GB
   vCPUs :    2
   Disksize:  20 GB
   Boot ISO:  https://github.com/minishift/minishift-b2d-iso/releases/download/v1.0.2/minishift-b2d.iso
-- Starting Minishift VM ... 
   Downloading ISO 'https://github.com/minishift/minishift-b2d-iso/releases/download/v1.0.2/minishift-b2d.iso'
 40.00 MiB / 40.00 MiB [==========================================] 100.00%- s
OK
-- Checking for IP ... OK
-- Checking host reachable from VM ... 
   Pinging 8.8.8.8 ... OK
-- Checking HTTP connectivity from VM ... 
   Retrieving http://minishift.io/index.html ... OK
-- Checking storage disk mounted ... OK
-- Checking storage disk usage ... 0% OK
-- Downloading OpenShift binary 'oc' version 'v1.5.1'
 18.94 MiB / 18.94 MiB [==========================================] 100.00% 0s
-- Checking `oc` support for startup flags ... 
   routing-suffix ... OK
   host-config-dir ... OK
   host-data-dir ... OK
   host-pv-dir ... OK
   host-volumes-dir ... OK
-- Checking OpenShift client ... OK
-- Checking Docker client ... OK
-- Checking Docker version ... OK
-- Checking for existing OpenShift container ... OK
-- Checking for openshift/origin:v1.5.1 image ... 
   Pulling image openshift/origin:v1.5.1
   Pulled 0/3 layers, 3% complete
   Pulled 0/3 layers, 74% complete
   Pulled 1/3 layers, 86% complete
   Pulled 2/3 layers, 99% complete
   Pulled 3/3 layers, 100% complete
   Extracting
   Image pull complete
-- Checking Docker daemon configuration ... OK
-- Checking for available ports ... OK
-- Checking type of volume mount ... 
   Using Docker shared volumes for OpenShift volumes
-- Creating host directories ... OK
-- Finding server IP ... 
   Using 192.168.64.2 as the server IP
-- Starting OpenShift container ... 
   Creating initial OpenShift configuration
   Starting OpenShift using container 'origin'
   Waiting for API server to start listening
   OpenShift server started
-- Adding default OAuthClient redirect URIs ... OK
-- Installing registry ... OK
-- Installing router ... OK
-- Importing image streams ... OK
-- Importing templates ... OK
-- Login to server ... OK
-- Creating initial project "myproject" ... OK
-- Removing temporary directory ... OK
-- Checking container networking ... OK
-- Server Information ... 
   OpenShift server started.
   The server is accessible via web console at:
       https://192.168.64.2:8443

   You are logged in as:
       User:     developer
       Password: developer

   To login as administrator:
       oc login -u system:admin
```

**Comment by gbraad** [comment]
Since v3.6.0 they have reduced the output when doing `oc cluster up`, which means that it feels like the deployment hangs between the point of `Image download complete` and `OpenShift server started`. :-s

**Comment by gbraad** [comment]
Fixes: #997, #966

**Comment by gbraad** [comment]
Restart will show the following:

```
-- Checking if KVM driver is installed ... 
   Driver is available at /usr/local/bin/docker-machine-driver-kvm ... OK
-- Starting local OpenShift cluster using 'kvm' hypervisor ...
-- Starting Minishift VM ... OK
-- Checking for IP address ... OK
-- Checking if external host is reachable from the Minishift VM ... 
   Pinging 8.8.8.8 ... OK
-- Checking HTTP connectivity from the VM ... 
   Retrieving http://minishift.io/index.html ... OK
-- Checking if persisten storage volume is mounted ... OK
-- Checking available disk space ... 0% OK
-- OpenShift cluster will be configured with ...
   Version: v1.5.1
-- Checking `oc` support for startup flags ... 
   host-config-dir ... OK
   host-data-dir ... OK
   host-pv-dir ... OK
   host-volumes-dir ... OK
   routing-suffix ... OK
-- Checking OpenShift client ... OK
-- Checking Docker client ... OK
-- Checking Docker version ... OK
-- Checking for existing OpenShift container ... OK
-- Checking for openshift/origin:v1.5.1 image ... 
   Pulling image openshift/origin:v1.5.1
...
```

Added feedback:
  * OpenShift version

**Comment by hferentschik** [comment]
Merged via pull request #1162

**Comment by hferentschik** [comment]
I also created #1219. I think we should re-visit the whole troubleshooting section and try to find a better format/structure for it.
